doc/whatsnew/fragments/6242.bugfix

```diff
@@ -0,0 +1,5 @@
+Use dict to track results while expanding the module files
+from list of module directories, package names, and standalone files to
+prevent duplicate scans while preserving order.
+
+Closes #6242
```

Comment:
```suggestion
Closes #6242, #4053
```
Very probably, and maybe also #689

Reply:
When I originally looked at #689, I thought that it was unrelated. But, I see that the originally-reported issue has been resolved and that the last comment in that issue thread _might_ be related, though I am not sure since I have not altered the package detection machinery or how `modname`/`basemodname` are reported. I will try the reproducer that was provided to see if this patch makes a difference.

I definitely agree that there is a high probability that this closes #4053.

Reply:
Updated Town Crier fragment.